tracer/events.go

```diff
@@ -178,8 +180,47 @@ func startPollingPerfEventMonitor(ctx context.Context, perfEventMap *ebpf.Map,
 					noDataCount.Add(1)
 					continue
 				}
-				triggerFunc(data.RawSample, data.CPU)
+
+				// Keep track of min KTime seen in this batch processing loop
+				trace := t.loadBpfTrace(data.RawSample, data.CPU)
+				if minKTime == 0 || trace.KTime < minKTime {
+					minKTime = trace.KTime
+				}
+				traceOutChan <- trace
+			}
+			// After we've received and processed all trace events, call
+			// SymbolizationComplete if there is a pending oldKTime that we
+			// haven't yet propagated to the rest of the agent.
+			//
+			// This introduces both an upper bound to SymbolizationComplete
+			// call frequency (dictated by pollTicker) but also skips calls
+			// when none are needed (e.g. no trace events have been read).
+			//
+			// We use oldKTime instead of minKTime (except when the latter is
+			// smaller than the former) to take into account scheduling delays
+			// that could in theory result in observed KTime going back in time.
+			//
+			// For example, as we don't control ordering of trace events being
+			// written by the kernel in per-CPU buffers across CPU cores, it's
+			// possible that given events generated on different cores with
+			// timestamps t0 < t1 < t2 < t3, this poll loop reads <t3 t1 t2>
+			// in a first iteration and t0 in a second iteration. If we use
+			// the current iteration minKTime we'll call
+			// SymbolizationComplete(t1) first and t0 next, with t0 < t1.
+			if oldKTime > 0 {
+				// Ensure that all previously sent trace events have been processed
+				traceOutChan <- nil
```

Comment:
1. This only ensures that the previous trace has been picked up by the reader, not that the trace has been processed already.
2. If the goal is to ensure that the previous trace has been processed, why not call HandleTrace() directly?

Reply:
It ensures that all previously sent trace events have been processed, otherwise `traceOutChan <- nil` would block as the channel is unbuffered. It's an explicit synchronization point. Calling `HandleTrace` here introduces direct coupling of `tracer` and `tracehandler`. 

Also it's trivial to introduce batching and parallelization (see [this](https://github.com/open-telemetry/opentelemetry-ebpf-profiler/pull/307#discussion_r1924470977) comment) and keep separation of concerns across subsystems with the existing architecture.